### DISCUSSION TITLE
Problems with private npm registry

### DISCUSSION THREAD

**Discussion by dbasilio** [renovatebot/renovate#9980] [Request Help]
Hello! We just upgraded to yarn 2 and are experimenting with setting up renovate for all of our repos. We're having issues with our private npm registry (hosted on jfrog). Here is the relevant part of the config we're using:

```json
{
    "npmrc": "registry=https://someUrl.jfrog.io/npm/",
    "hostRules": [
        {
            "hostType": "npm",
            "baseUrl": "https://someUrl.jfrog.io/npm/",
            "authType": "Basic",
            "username": "***redacted***",
            "encrypted": {
                "password": "***redacted****"
            }
        }
    ],
    "postUpdateOptions": [
        "yarnDedupeHighest"
    ],
}
```

Using this config, I saw from the onboarding PR that renovate was able to find our internal packages. We merged the onboarding PR, but all the PRs generated have errors. We run `yarn install --immutable` in CI as no changes should be generated there, so the PRs are all failing CI as well. Renovate fails with:

![image](https://user-images.githubusercontent.com/8311284/118013212-e149cb80-b31f-11eb-8888-c1ada551d70d.png)

When I dig into the logs I see: 

```json
{ 
    "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/our-app-name\":\"/mnt/renovate/gh/our-app-name\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_ENABLE_IMMUTABLE_INSTALLS -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/mnt/renovate/gh/our-app-name\" docker.io/renovate/node:14.15.5 bash -l -c \"npm i -g yarn && yarn install && yarn dedupe --strategy highest\"",
    "stdout": "\n> yarn@1.22.10 preinstall /home/ubuntu/.npm-global/lib/node_modules/yarn\n> :; (node ./preinstall.js > /dev/null 2>&1 || true)\n\n/home/ubuntu/.npm-global/bin/yarn -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n/home/ubuntu/.npm-global/bin/yarnpkg -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n+ yarn@1.22.10\nadded 1 package in 1.605s\n➤ YN0000: ┌ Resolution step\n➤ YN0035: │ our-package@npm:^2.1.67: The remote server failed to provide the requested resource\n➤ YN0035: │   Response Code: 404 (Not Found)\n➤ YN0035: │   Request Method: GET\n➤ YN0035: │   Request URL: https://**redacted**our-package\n➤ YN0000: └ Completed in 1s 227ms\n➤ YN0000: Failed with errors in 1s 240ms\n" 
}
```
So it seems like the credentials for the private npm modules are not being passed into the container for the `yarn install` command to be run with. Wondering if there's something I'm clearly doing wrong here?

### Why am I using  npmrc and also host rules?
Without the `npmrc`, the requests were going to 'register.npmjs.org', even with the `hostRules` as it is shown above.

### Why am I using username/password instead of `npmToken`?
The access token generated by jfrog is too long for the encrypt page, so we have no choice but to use username/password

**Comment by rarkins** [comment]
Can you share your `.yarnrc.yml`? e.g. I want to see what info it has in it regarding scopes/registries

I think the problem is:
- Renovate should read scope/registry mappings from `.yarnrc.yml` so that it looks up from the correct registry internally (not needing npmrc to be configured)
- Renovate should take auth from hotsRules and add/append it to the existing `.yarnrc.yml` before running `yarn install`, so that it has credentials for the registry

**Reply by rarkins** [reply]
Let's move the discussion to #9985
